Posts/2021/10/tfl-secrets.md

```diff
@@ -0,0 +1,228 @@
+                                                         Set-SecretStorePassword], [Get-SecretStoreConfiguration,
+                                                         Get-SecretStoreConfiguration], [Set-SecretStoreConfiguration,
+                                                         Set-SecretStoreConfiguration], [Reset-SecretStore,
+                                                         Reset-SecretStore]}
+
+```
+
+As you can see, both modules have a number of commands you may need to use to manage secrets for your environment.
+
+## Registering and viewing a secret vault
+
+After you have the two modules installed, your next step is to register a secret vault.
+There are several vault options you can take advantage of, for this post, I'll use the built-in default vault.
+You configure the default vault like this:
+
+```powershell-console
+PS C:\Foo> # 4. Register the default secrets provider
+PS C:\Foo> $Mod = 'Microsoft.PowerShell.SecretStore'
+PS C:\Foo> Register-SecretVault -Name RKSecrets -ModuleName $Mod -DefaultVault
+PS C:\Foo> Get-SecretVault
+
+Name      ModuleName                       IsDefaultVault
+----      ----------                       --------------
+RKSecrets Microsoft.PowerShell.SecretStore True
+```
+Like the previous step, registering the vault does not create any output by default.
+You can view the vault you just created by using the `Get-SecretVault` command.
+
+## Setting a secret
+
+To create a new secret in your secret vault, you use the 1Set-Secret1 command, like this:
+
+```powershell-console
+
+PS> # 4. Register the default secrets provider
+PS> $Mod = 'Microsoft.PowerShell.SecretStore'
+PS> Register-SecretVault -Name RKSecrets -ModuleName $Mod -DefaultVault
+PS> # 5. View Secret vault
+PS> Get-SecretVault
+
+Name      ModuleName                       IsDefaultVault
+----      ----------                       --------------
+RKSecrets Microsoft.PowerShell.SecretStore True
+
+PS C:\Foo> # 6. Set the Admin password secret for Reskit forest
+PS C:\Foo> Set-Secret -Name ReskitAdmin -Secret 'Pa$$w0rd'
+Creating a new RKSecrets vault. A password is required by the current store configuration.
+Enter password:
+**********
+Enter password again for verification:
+**********
+```
+
+The first time you use `Set-Secret` to create a secret, the cmdlet prompts for a vault password.
+Note this password isd NOT stored in the AD - so don't forget it!!!
+
+## Using secrets stored in your secret vault
+
+Now that you have set a password in the RKSecrets vault, you can use the `Get-Secret` cmdlet to retrieve the secret.
+As you can see here, although you set a plain text password, Get-Secret returns the secret as a secure string.
+
+```powershell-console
+PS> # 7. Create a credential object using the secet
+PS> $User = 'Reskit\Administrator'
+PS> $PwSS = Get-Secret ReskitAdmin
+PS> $Cred = [System.Management.Automation.PSCredential]::New($User,$PwSS)
```

Comment:
For folks new to these concepts they may not understand what you are doing here. The goal is to convert the username/password pair into a PSCredential object, because that is what the PowerShell cmdlets want to use. And you might want to explain why you need to use the .NET class for this. I know that `Get-Credential` does not allow you to pass in a password, but many people may not.